src/msw/webrequest_winhttp.cpp

```diff
@@ -12,23 +12,121 @@
+    static WinHttpSetOption_t WinHttpSetOption;
+    typedef BOOL(WINAPI* WinHttpWriteData_t)(HINTERNET, LPCVOID, DWORD, LPDWORD);
+    static WinHttpWriteData_t WinHttpWriteData;
+    typedef BOOL(WINAPI* WinHttpCloseHandle_t)(HINTERNET);
+    static WinHttpCloseHandle_t WinHttpCloseHandle;
+    typedef BOOL(WINAPI* WinHttpReceiveResponse_t)(HINTERNET, LPVOID);
+    static WinHttpReceiveResponse_t WinHttpReceiveResponse;
+    typedef BOOL(WINAPI* WinHttpCrackUrl_t)(LPCWSTR, DWORD, DWORD, LPURL_COMPONENTS);
+    static WinHttpCrackUrl_t WinHttpCrackUrl;
+    typedef HINTERNET(WINAPI* WinHttpConnect_t)(HINTERNET, LPCWSTR, INTERNET_PORT, DWORD);
+    static WinHttpConnect_t WinHttpConnect;
+    typedef HINTERNET(WINAPI* WinHttpOpenRequest_t)(HINTERNET, LPCWSTR, LPCWSTR, LPCWSTR, LPCWSTR, LPCWSTR*, DWORD);
+    static WinHttpOpenRequest_t WinHttpOpenRequest;
+    typedef WINHTTP_STATUS_CALLBACK(WINAPI* WinHttpSetStatusCallback_t)(HINTERNET, WINHTTP_STATUS_CALLBACK, DWORD, DWORD_PTR);
+    static WinHttpSetStatusCallback_t WinHttpSetStatusCallback;
+    typedef BOOL(WINAPI* WinHttpSendRequest_t)(HINTERNET, LPCWSTR, DWORD, LPVOID, DWORD, DWORD, DWORD_PTR);
+    static WinHttpSendRequest_t WinHttpSendRequest;
+    typedef BOOL(WINAPI* WinHttpReadData_t)(HINTERNET, LPVOID, DWORD, LPDWORD);
+    static WinHttpReadData_t WinHttpReadData;
+    typedef BOOL(WINAPI* WinHttpQueryAuthSchemes_t)(HINTERNET, LPDWORD, LPDWORD, LPDWORD);
+    static WinHttpQueryAuthSchemes_t WinHttpQueryAuthSchemes;
+    typedef BOOL(WINAPI* WinHttpSetCredentials_t)(HINTERNET, DWORD, DWORD, LPCWSTR, LPCWSTR, LPVOID);
+    static WinHttpSetCredentials_t WinHttpSetCredentials;
+    typedef HINTERNET(WINAPI* WinHttpOpen_t)(LPCWSTR, DWORD, LPCWSTR, LPCWSTR, DWORD);
+    static WinHttpOpen_t WinHttpOpen;
+
+private:
+    static wxDynamicLibrary m_winhttp;
+};
+
+wxDynamicLibrary wxWinHTTP::m_winhttp;
+wxWinHTTP::WinHttpQueryOption_t wxWinHTTP::WinHttpQueryOption;
+wxWinHTTP::WinHttpQueryHeaders_t wxWinHTTP::WinHttpQueryHeaders;
+wxWinHTTP::WinHttpSetOption_t wxWinHTTP::WinHttpSetOption;
+wxWinHTTP::WinHttpWriteData_t wxWinHTTP::WinHttpWriteData;
+wxWinHTTP::WinHttpCloseHandle_t wxWinHTTP::WinHttpCloseHandle;
+wxWinHTTP::WinHttpReceiveResponse_t wxWinHTTP::WinHttpReceiveResponse;
+wxWinHTTP::WinHttpCrackUrl_t wxWinHTTP::WinHttpCrackUrl;
+wxWinHTTP::WinHttpConnect_t wxWinHTTP::WinHttpConnect;
+wxWinHTTP::WinHttpOpenRequest_t wxWinHTTP::WinHttpOpenRequest;
+wxWinHTTP::WinHttpSetStatusCallback_t wxWinHTTP::WinHttpSetStatusCallback;
+wxWinHTTP::WinHttpSendRequest_t wxWinHTTP::WinHttpSendRequest;
+wxWinHTTP::WinHttpReadData_t wxWinHTTP::WinHttpReadData;
+wxWinHTTP::WinHttpQueryAuthSchemes_t wxWinHTTP::WinHttpQueryAuthSchemes;
+wxWinHTTP::WinHttpSetCredentials_t wxWinHTTP::WinHttpSetCredentials;
+wxWinHTTP::WinHttpOpen_t wxWinHTTP::WinHttpOpen;
+
+class wxWinHTTPModule : public wxModule
+{
+public:
+    virtual bool OnInit() wxOVERRIDE
+    {
+        return wxWinHTTP::LoadLibraries();
```

Comment:
I don't know if this is going to really be a problem in practice, but failing to load `winhttp.dll`, for whatever reason, will now prevent any wx application (linking with wxnet) from starting up, even if it doesn't use `wxWebRequest` at all. Again, this might well never happen, but it still seems a bit worrisome.

It would be better to do this on demand only, but I'm not sure if we know which function will always be called first. Can we rely on no WinHTTP functions being called before `wxWebRequestWinHTTP` ctor?

Also, there is still a risk of crashing if we fail and the function pointer is null.

The only robust solution I see is to make all the function wrappers, i.e. check if it could be loaded and return an error otherwise, but this would require even more boilerplate code.

So, finally, the best solution I see is to do this from `wxWebSession::InitFactoryMap()`, i.e. add a check for successful initialization there before creating `wxWebSessionFactoryWinHTTP`. This should keep things reasonably simple, while ensuring that we don't do anything worse than not allowing to use WinHTTP if we fail to load the DLL or resolve any functions in it.

What do you think?

Reply:
I agree, I'll move it to the initialization of `wxWebSessionFactoryWinHTTP` and check for success.